src/NuGet.Clients/NuGet.VisualStudio.Common/VisualStudioBuildItemStorageCallback.cs

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace NuGet.VisualStudio
+{
+    /// <summary>
+    /// <see cref="VisualStudioBuildItemStorageCallback" it is used together with <see cref="IVsBuildItemStorage"/> to find a project's items and read their metadata.
+    /// </summary>
+    public class VisualStudioBuildItemStorageCallback : IVsBuildItemStorageCallback
+    {
+        /// <summary>
+        /// The list of Ids and requested metadata.
+        /// For a project item like
+        /// <![CDATA[<PackageVersion Include="Foo" Version="1.1.1"/>]]>
+        /// If the requested metadata is "Version" the returned Items will be (Foo, {"1.1.1"})
+        /// </summary>
+        public List<(string Itemid, List<string> ItemMetadata)> Items { get; } = new List<(string Itemid, List<string> ItemMetadata)>();
+
+        private VisualStudioBuildItemStorageCallback()
+        {
+        }
+
+        public static VisualStudioBuildItemStorageCallback Instance => new VisualStudioBuildItemStorageCallback();
```

Comment:
Static instances only work for stateless or immutable objects. This object's `Items` property gets modified, which means if you have multiple legacy PR projects, or you unload and reload a single legacy PR project, bad things will happen.

Reply:
Discussed offline. There is not an issue with the functionality. However it is hard to read, so changed I refactored the code.